diag/diagnostic.go

```diff
@@ -0,0 +1,49 @@
+package diag
+
+import (
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+// Diagnostic is an interface for providing enhanced feedback.
+//
+// These are typically practitioner facing, however it is possible for
+// functionality, such as validation, to use these to change behaviors or
+// otherwise have these be manipulated or removed before being presented.
+//
+// See the ErrorDiagnostic and WarningDiagnostic concrete types for generic
+// implementations.
+type Diagnostic interface {
+	// Severity returns the desired level of feedback for the diagnostic.
+	Severity() Severity
+
+	// Summary is a short description for the diagnostic.
+	//
+	// Typically this is implemented as a title, such as "Example Error",
```

Comment:
Would it be possible to use a real-world example here? Like "Invalid resource name" or something? I think it may be more illustrative.